// @ngforage/ngforage-ng5/dist/esm5/NgForage/instance-factory/InstanceFactory.service.js
import { Injectable } from '@angular/core';
import * as lf from 'localforage';
/** @internal */
var instance;
/** @internal */
var stores = {};
/** @internal */
function getDriverString(driver) {
    if (!driver) {
        return '';
    }
    else if (typeof driver === 'string') {
        return driver;
    }
    else {
        return driver.slice().sort().join(',');
    }
}
/** @internal */
function getHash(cfg) {
    return [
        getDriverString(cfg.driver),
        cfg.name,
        cfg.size,
        cfg.storeName,
        cfg.version,
        cfg.description,
        cfg.cacheTime
    ].join('|');
}
/**
 * Creates localForage instances
 */
var InstanceFactory = (function () {
    function InstanceFactory() {
    }
    InstanceFactory.factory = function () {
        if (!instance) {
            instance = new InstanceFactory();
        }
        return instance;
    };
    /** @internal */
    /** @internal */
    InstanceFactory.prototype.getInstance = /** @internal */
    function (cfg) {
        var hash = getHash(cfg);
        if (!stores[hash]) {
            stores[hash] = lf.createInstance(cfg);
        }
        return stores[hash];
    };
    /** @internal */
    InstanceFactory.provider = {
        provide: InstanceFactory,
        useFactory: InstanceFactory.factory
    };
    InstanceFactory.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    InstanceFactory.ctorParameters = function () { return []; };
    return InstanceFactory;
}());
export { InstanceFactory };
